pkg/transport/generic_transport_controller.go

```diff
@@ -0,0 +1,512 @@
+	}
+
+	wrappedObject, err := convertObjectToUnstructured(c.transport.WrapObjects(objectsToPropagate))
+	if err != nil {
+		return fmt.Errorf("failed to wrap objects to a single wrapped object - %w", err)
+	}
+	// wrapped object name is (PlacementDecision.GetName()-WdsName).
+	// pay attention - we cannot use the PlacementDecision object name, cause we might have duplicate names coming from different WDS spaces.
+	// we add WdsName to the object name to assure name uniqueness,
+	// in order to easily get the origin PlacementDecision object name and wds, we add it as an annotations.
+	wrappedObject.SetName(fmt.Sprintf("%s-%s", placementDecision.GetName(), c.wdsName))
+	setAnnotation(wrappedObject, originOwnerReferenceAnnotation, placementDecision.GetName())
+	setAnnotation(wrappedObject, originWdsAnnotation, c.wdsName)
+
+	for _, destination := range placementDecision.Spec.Destinations { // TODO need to revisit the destination struct and see how to use it properly
+		if err := c.createOrUpdateWrappedObject(ctx, destination.ClusterId, wrappedObject); err != nil {
+			return fmt.Errorf("failed to propagate wrapped object '%s' to all required WECs - %w", wrappedObject.GetName(), err)
+		}
+	}
+
+	return nil
+}
+
+func (c *genericTransportController) getObjectsFromWDS(ctx context.Context, placementDecision *v1alpha1.PlacementDecision) ([]*unstructured.Unstructured, error) {
+	objectsToPropagate := make([]*unstructured.Unstructured, 0)
+	// add cluster-scoped objects to the 'objectsToPropagate' slice
+	for _, clusterScopedObject := range placementDecision.Spec.Workload.ClusterScope {
+		if clusterScopedObject.ObjectNames == nil {
+			continue // no objects from this gvr, skip
+		}
+		gvr := schema.GroupVersionResource{Group: clusterScopedObject.Group, Version: clusterScopedObject.Version, Resource: clusterScopedObject.Resource}
+		for _, objectName := range clusterScopedObject.ObjectNames {
+			object, err := c.wdsDynamicClient.Resource(gvr).Get(ctx, objectName, metav1.GetOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("failed to get required cluster-scoped object '%s' with gvr %s from WDS - %w", objectName, gvr, err)
+			}
+			objectsToPropagate = append(objectsToPropagate, cleanObject(object))
+		}
+	}
+	// add namespace-scoped objects to the 'objectsToPropagate' slice
+	for _, namespaceScopedObject := range placementDecision.Spec.Workload.NamespaceScope {
+		gvr := schema.GroupVersionResource{Group: namespaceScopedObject.Group, Version: namespaceScopedObject.Version, Resource: namespaceScopedObject.Resource}
+		for _, objectsByNamespace := range namespaceScopedObject.ObjectsByNamespace {
+			if objectsByNamespace.Names == nil {
+				continue // no objects from this namespace, skip
+			}
+			for _, objectName := range objectsByNamespace.Names {
+				object, err := c.wdsDynamicClient.Resource(gvr).Namespace(objectsByNamespace.Namespace).Get(ctx, objectName, metav1.GetOptions{})
```

Comment:
`c.wdsDynamicClient.Resource(gvr)` could be computed once, before the outer loop; `.Namespace(objectsByNamespace.Namespace)` could be done once, before the inner loop.

Reply:
It had better be allocating some new memory, otherwise two calls with different namespaces (which could happen concurrently) will interfere with each other!
Not a blocker.

Reply:
you're right. it does allocate some new memory.
my mistake.
will take care in later PR